design/vela-core/environment_template_appgroup.md

```diff
@@ -0,0 +1,132 @@
+
+- Users have a template base, and want to make modifications based on selected environment. For example, set different image names for dev/prod environments.
+- Users want the definitions in dev/prod environments to be different. For example, the logging trait might be uploaded to OSS/S3 buckets in dev or sinked to ELK in prod.
+- Users want the secrets in dev/prod environments to be different. For example, the DB connection secrets are different in dev/prod environments.
+
+
+## Proposal
+
+This design proposes to add the following CRDs to satisfy user requirements.
+
+### 1 Environment CRD
+
+Environment CRD defines the shared-base environment to deploy applications to.
+
+```yaml
+kind: Environment
+spec:
+  clusters: # Apply to the following clusters.
+    - prod-cluster
+
+  secrets: # The secrets that will be created/updated.
+    - name: redis
+      data:
+        url: redis-url
+        password: redis-password
+
+  definitions: # The definitions that will be created/updated.
+    - type: Component
+      name: function
+      source:
+        kube: # using kubectl apply 
+          git: git-repo-url
+          path: path/to/definition # path to a file or a directory
+    - type: Trait
+      name: logging
+      source:
+        helm: # using helm install/upgrade
+          git: git-repo-url
+          path: path/to/chart
+          valuePath: value.yaml
+```
+
+We would add a new env-controller to handle this CR. When a new Environment CR is applied, the env-controller will apply the secrets and definitions to the corresponding clusters. Note that the cluster objects are defined in [v1beta1/cluster_types.go](https://github.com/oam-dev/kubevela/blob/675b0e24db0cedde11c9870242763957b0012f99/apis/core.oam.dev/v1beta1/cluster_types.go#L45-L52).
+
+### 2. ApplicationTemplate CRD
+
+The ApplicationTemplate CRD defines the base template shared to further parameterize or patch more config to.
+
+```yaml
+kind: ApplicationTemplate
+spec:
+  # base template written in CUE
+  template: |
+    output: {
+      apiVersion: core.oam.dev/v1alpha2
+      kind: Application
+      metadata:
+        name: context.name
+      spec:
+        components:
+          - name: instance
+            type: webservice
+            properties:
+              image: test-image
+              cmd:
+                - sleep 100
+            traits:
+              - name: ingress
+          - name: database
+            type: parameter.databaseType
+        environment:
+          namespace: parameter.project
+    }
+    parameter: {
+      project: string | *""
+      databaseType: string | *"rds"
+    }
+```
+
+Note that this is a data object to reference to. It wouldn't have any effects if applied by itself.
+
+### 3. ApplicationSetting CRD
+
+With the above ApplicationTemplate, how could users pick a template and define environment-specific parameters? This is what ApplicationSetting CRD defines.
+
+```yaml
+kind: ApplicationSetting
+spec:
+  applications:
```

Comment:
Are we able to make `Application` the entry point? We should avoid having too many entries in general.

Reply:
In my opinion, `ApplicationTemplate`  is a top-level object. Users could directly create `Application` object, or use `ApplicationTemplate`+`ApplicationSettings` to generate `Application` in various environments.  

It may be strange to somebody if  `ApplicationTemplate` as a policy of `Application`, at least for me.

@hongchaodeng @resouer

Reply:
> It is not about making ApplicationTemplate as a policy. It is about making the templating method as a policy, e.g. cue/helm/kustomize. We can still have the upper layer like ApplicationSet to manage config.

@hongchaodeng  Got it.

Reply:
Adding a note here. As I'm implementing the workflow, I realize that this is not intuitive in workflow. The config templating or overlay could be better done at admission layer.